lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+// It also writes the offset of the entry in the file to a separate index file, such that a consumer can look up a
+// particular entry index in the state change file.
+func (stateChangeSyncer *StateChangeSyncer) _handleStateSyncerOperation(event *StateSyncerOperationEvent) {
+	// If we're in blocksync mode, we only want to flush entries once the sync is complete.
+	if !stateChangeSyncer.BlocksyncCompleteEntriesFlushed && stateChangeSyncer.SyncType == NodeSyncTypeBlockSync {
+		return
+	}
+
+	stateChangeEntry := event.StateChangeEntry
+
+	// Check to see if the index in question has a "core_state" annotation in its definition.
+	if !isCoreStateKey(stateChangeEntry.KeyBytes) {
+		return
+	}
+
+	// Make sure 2 operations aren't logged simultaneously.
+	stateChangeSyncer.StateSyncerMutex.Lock()
+	defer stateChangeSyncer.StateSyncerMutex.Unlock()
+
+	flushId := event.FlushId
+
+	// Crate a block sync flush ID if one doesn't already exist.
+	if event.FlushId == uuid.Nil && stateChangeSyncer.BlockSyncFlushId == uuid.Nil {
+		stateChangeSyncer.BlockSyncFlushId = uuid.New()
+	}
+
+	if event.IsMempoolTxn {
+		// Create a mempool flush ID if one doesn't already exist.
+		if event.FlushId == uuid.Nil && stateChangeSyncer.MempoolFlushId == uuid.Nil {
+			stateChangeSyncer.MempoolFlushId = uuid.New()
+		}
+
+		// If the event flush ID is nil, then we need to use the global mempool flush ID.
+		if event.FlushId == uuid.Nil {
+			flushId = stateChangeSyncer.MempoolFlushId
+		}
+
+		// The current state of the tracked mempool is stored in the MempoolKeyValueMap. If this entry is already in there
+		// then we don't need to re-write it to the state change file.
+		// Create key for op + key map
+		txKey := fmt.Sprintf("%v%v", event.StateChangeEntry.OperationType, string(event.StateChangeEntry.KeyBytes))
+
+		// Check to see if the key is in the map, and if the value is the same as the value in the event.
+		if valueBytes, ok := stateChangeSyncer.MempoolKeyValueMap[txKey]; ok && bytes.Equal(valueBytes, event.StateChangeEntry.EncoderBytes) {
+			// If the key is in the map, and the entry bytes are the same as those that are already tracked by state syncer,
+			// then we don't need to write the state change entry to the state change file - it's already being tracked.
+			return
+		}
+
+		// Add or remove from our tracking maps based on operation types.
+		if stateChangeEntry.OperationType == DbOperationTypeDelete {
+			delete(stateChangeSyncer.MempoolKeyValueMap, txKey)
+			delete(stateChangeSyncer.MempoolFlushKeySet, txKey)
+		} else {
+			stateChangeSyncer.MempoolKeyValueMap[txKey] = event.StateChangeEntry.EncoderBytes
+			stateChangeSyncer.MempoolFlushKeySet[txKey] = true
+		}
+	} else {
+		// If the flush ID is nil, then we need to use the global block sync flush ID.
+		if event.FlushId == uuid.Nil {
+			flushId = stateChangeSyncer.BlockSyncFlushId
+		}
+	}
+
+	// Get the relevant deso encoder for this keyBytes.
+	var encoderType EncoderType
```

Comment:
More comments here. I want to better understand when something hits the first case vs the second. I think it happens whenever we have a field in the db that's a "key only" field with no Entry as a value.